internal/blocksync/reactor.go

```diff
@@ -442,6 +449,17 @@ FOR_LOOP:
 				panic(fmt.Errorf("peeked first block without extended commit at height %d - possible node store corruption", first.Height))
 			}
 
+			// Before priming didProcessCh for another check on the next
+			// iteration, break the loop if the BlockPool or the Reactor itself
+			// has quit. This avoids case ambiguity of the outer select when two
+			// channels are ready.
+			select {
+			case <-bcR.pool.Quit():
+				break FOR_LOOP
+			case <-bcR.Quit():
+				break FOR_LOOP
+			default:
+			}
```

Comment:
The same effect is achieved with an if testing if the two services are running (`IsRunning()`). With the advantage that the "atomic bool" used by this call is updated before closing the quit channel.

Reply:
Moreover, this is not really relevant for the solution.

The panic happens because this reactor is stopped, but this routine is not. If you solve this issue, you shouldn't have the panics coming from the block executor and the associated databases.

The reason for which is that when stopping the node (`Node.Stop()`) we first stop all reactors (the `Switch.OnStop()` does that), to then stop the other components. So while this reactor has not being stopped, which only happens when this routine leaves, we should not observe the reported problem.

Reply:
> The same effect is achieved with an if testing if the two services are running (`IsRunning()`). With the advantage that the "atomic bool" used by this call is updated before closing the quit channel.

Ah, yes, I see the CAS in `(*BaseService).Stop` in advance of closing the quit channel.  Good call.  No need for a select in this spot unlike the outer `select`.

> The panic happens because this reactor is stopped, but this routine is not. If you solve this issue, you shouldn't have the panics coming from the block executor and the associated databases.
> 
> The reason for which is that when stopping the node (Node.Stop()) we first stop all reactors (the Switch.OnStop() does that), to then stop the other components. So while this reactor has not being stopped, which only happens when this routine leaves, we should not observe the reported problem.

Yes, absolutely correct that this is not necessary to synchronize the `poolRoutine` return with the service shutdown.  I just noticed the outer select was in a contest due to the subsequent `didProcessCh` being primed for the "Trying again quickly".

Waiting on `poolRoutine` did indeed solve the issue, but there were often many iterations of the loop after stop due to this quick retry path not being escaped by shutdown.

Thanks for the insightful comments here!

Reply:
I think you still have to check the `bcR.pool` also here for having the purposed effect.

The logic, as it appears is: stop the reactor, `OnStop` stops the `pool`, then waits for this routine to leave.

Reply:
>  I just noticed the outer select was in a contest due to the subsequent didProcessCh being primed for the "Trying again quickly". That is likely unintended.

This is true and, to be honest, a very debatable approach.